### ISSUE TITLE
[TODO]: `SPOOF_PROTECTION=1` feature and Alias support improvements

### ISSUE THREAD

**Issue by polarathene** [docker-mailserver/docker-mailserver#2819]
Adding this here in case anyone else has time to tackle any of it.

I am not likely to have the time until at least December, but I expect I'll have to prioritize my time on other commitments before I can come back to addressing these tasks.

## Task Overview

I've provided plenty of helpful details below, but a summary of tasks to complete is:

**`SPOOF_PROTECTION=1` feature:**

- [ ] [Only bother with the `unionmap` and drop the conditional](https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/scripts/startup/setup-stack.sh#L471-L476), as the file always exists, there is no known reason to need the alternative without unionmap or additional regex table.
- [ ] In addition, add support for another table to that `unionmap`, which is not a regex table. It will allow explicitly allowing additional senders for an account to send as.
  - [ ] Once supported, update the docs workaround example covered in `@gmail` related issue section below.
- [ ] Consider if `unionmap` would also be relevant for the `ldap` equivalent value in that same linked method.
- [ ] Consider improving user docs (or eventual maintainer docs) to better present the relevant information for this feature covered below. Potentially provide an example with the `noreply@external-domain` from referenced send-only related issue below. Likewise with the LDAP specific config ENV that was added to support the feature specifically there.
- [ ] Review the mentioned/linked concerns below for enabling the feature by default in a future release, and decide if it should be default once better docs and config support is merged. The ENV name may also need reconsideration (also discussed below).

**Related alias support:**
- [ ]  `setup alias add` section below notes a bug to resolve, and CLI help message to update. It also has poor error handling feedback if some `setup` commands do not receive the expected number of inputs IIRC (_may have been `setup alias del`_).
- [ ] In the final section below _"Alias domains being added to `/etc/postfix/vhost`"_, it has been identified that we could improve how we determine what domains are managed.
  - Alias domains should be handled better and this support should be easier to exclude domains via using Postfix `main.cf` settings more appropriately.
  - Keep in mind `/etc/postfix/vhost` is presently sourced elsewhere, such as the DKIM generation support in `setup` command.

---

## `SPOOF_PROTECTION=1` handling for `smtpd_sender_login_maps`

Jan 2021, @georglauterbach [created and merged a PR](https://github.com/docker-mailserver/docker-mailserver/pull/1754) to fix a user reported bug. The PR introduced a [`unionmap: { ... }` table](https://www.postfix.org/DATABASE_README.html), which Postfix queries each listed table and merges the responses, rather than accepting the first matched response and exiting early with that outcome (_as the result may fail, but have passed from another valid match rule later on_).

Nothing wrong there, but it was the PR that split the non-LDAP handling for [`smtpd_sender_login_maps`](https://www.postfix.org/postconf.5.html#smtpd_sender_login_maps) (_used for ENV `SPOOF_PROTECTION=1`_) into two values, only allowing the `unionmap` when `/etc/postfix/regexp` config was present. This does not seem necessary, and using the `unionmap` regardless is probably fine?

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/scripts/helpers/aliases.sh#L29-L37

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/scripts/start-mailserver.sh#L116-L123

`/etc/postfix/regexp` is always created during startup, and it has always been before this logic is run, even before that PR which added the conditional check on it's existence. We would always get `unionmap` AFAIK. It may even be useful for the LDAP configs.

---

As a refresher for anyone reading `SPOOF_PROTECTION=1` (opt-in) prevents submitting any mail to send _unless_ the logged in account is authorized to send from that envelope address (`FROM`) in `smtpd_sender_login_maps` lookup tables. The ENV does this by adding this restriction: `smtpd_sender_restrictions = reject_authenticated_sender_login_mismatch` (_prefixed in front of other sender restrictions_).

Without this, our default config allows a logged in user to send mail "from" whatever address they'd like to, including other accounts for the same domain, or one that DMS doesn't manage. The related bug report that introduced the fix with `unionmap`, has a user [explain a good use-case for wanting that](https://github.com/docker-mailserver/docker-mailserver/issues/1437#issuecomment-613378211), non-human service accounts being restricted to their account only for sending mail in-case the service is compromised, while using a regex to allow their `mail@example.com` account (admin) to be able to send as any address for that domain.

When opting into `SPOOF_PROTECTION=1`, we permit [local aliases (_`root`, `amavis`, etc_)](https://github.com/docker-mailserver/docker-mailserver/pull/952) (_that linked PR introduced the `POSTMASTER_ADDRESS` ENV and alias, along with setting `mydestination` in `main.cf`_), account aliases, and a few other lookup tables in the `unionmap`. One of these is a regexp file [`sender_login_maps.pcre`](https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/postfix/sender_login_maps.pcre) that will match whatever the owners sender address is, and allow it (_There is similar advice elsewhere](https://serverfault.com/questions/318334/how-to-enforce-sender-address-to-be-logged-in-userexample-org-in-postfix/710235#710235), that suggests this approach, seems fine_). A more specific match would have higher priority I think, but it may just depend on whatever is matched first from top to bottom of a table.

It should be noted that without the `/etc/postfix/virtual` table for account aliases included in the `unionmap`, you cannot send from an alias (eg: `marketing@example.com`), but you can authenticate over SASL as your aliased account by using that alias as the login account/username. That still logs the user in as the aliased account, not the alias, thus the actual account is still what is checked for allowed sender addresses that can be used regardless.

The [original PR that introduced this feature with `SPOOF_PROTECTION=1`](https://github.com/docker-mailserver/docker-mailserver/pull/872#issuecomment-370242684) also has discussion about enabling it by default in future breaking change. And some users sharing concerns about their workflows being broken, one from a previous maintainer who was using aliases to reply to mail keeping the same `FROM` address in the message, while only having a single mail account to manage.

---

## Additional Notes

The [`/etc/postfix/regexp` table was added in this PR](https://github.com/docker-mailserver/docker-mailserver/pull/1032), when the user previously tried to use a catch-all alias `@example.com` and noticed it did not work for them (_received but could not send_), possibly related to the later fix of using a `unionmap` table.

If we have any LDAP related issues with the feature, [this may be helpful](https://github.com/docker-mailserver/docker-mailserver/issues/892#issuecomment-374161847). That had a [follow-up issue](https://github.com/docker-mailserver/docker-mailserver/issues/1340#issuecomment-566180500) which also had a [related PR merged](https://github.com/docker-mailserver/docker-mailserver/pull/1902) for improving LDAP support with an optional filter config ENV - it demonstrates how to allow an additional sender address for `noreply@example.com`.

---

The tables for `smtpd_sender_login_maps` are pretty simple, the left-side (1st column) is to match the desired sender address (_explicit value or regex pattern depending on table type_), and on the right-side (2nd column) is usually the owner address (_login mail account, non-aliased_), but it can be multiple addresses to allow too by using a `, ` delimiter just like the `/etc/postfix/virtual` alias table allows to map an alias to multiple recipients.

If using a regex pattern, keep in mind that whatever matches in a table first is the pass/fail for that table, regardless of the `unionmap` usage. As noted by [this users comment](https://github.com/Mailu/Mailu/issues/1096#issuecomment-632121639) that had two lines `/.*/` to approve two separate domains, only the first would ever be matched. The response suggests moving the 2nd line to a separate table, but they should be able to fold the 2nd domain into a single line for that same regexp as mentioned above with `, ` delimiter.

The [ENV docs for `SPOOF_PROTECTION=1`](https://docker-mailserver.github.io/docker-mailserver/edge/config/environment/#spoof_protection) notes that extension delimiters (eg: `+`) are no longer usable as we don't have a table to authorize them out of the box. Technically it should be possible with regex as well? Would need some testing to ensure it's not abused.

`unionmap` with `smtpd_sender_login_maps` is covered well with the problem it solves in [this blog article](https://docs.sweeting.me/s/using-multiple-smtpd-sender-login-maps-in-postfix). 

The [Postfix SASL docs](http://www.postfix.org/SASL_README.html#server_sasl_authz_envelope) briefly touch on related configuration discussed here, but seems to incorrectly reference using it with `smtpd_recipient_restrictions` rather than [`smtpd_sender_restrictions`](http://www.postfix.org/postconf.5.html#smtpd_sender_restrictions).

---

### `SPOOF_PROTECTION=1` only restricts the envelope `FROM` sender address, spoofing still feasible?

Despite `SPOOF_PROTECTION=1` applying these settings to restrict what a sender can use in `FROM` (mail envelope), this doesn't prevent spoofing of the `from:` (mail message) header apparently. This is mentioned [here](https://serverfault.com/questions/872932/postfix-smtpd-sender-login-maps-regex-matches-everything-instead-of-only-sasl/873149#873149), in a link it references, that resource [links to another](https://security.stackexchange.com/questions/30732/why-is-it-even-possible-to-forge-sender-header-in-e-mail/30741#30741) that clarifies the difference between envelope and message `FROM` headers, and how SPF+DKIM+DMARC are leveraged to better handle this.

Thus `SPOOF_PROTECTION` may not be the most appropriate ENV name, it's mostly to restrict your own authenticated accounts to what sender addresses can be used in the envelope `FROM` when they submit mail, which is still worthwhile to do. There is [this project](https://github.com/magcks/milterfrom) which is a Postfix milter designed to ensure the sender envelope and message are the same, which prevents OpenDKIM signing the message sender (`From:` header) for a mismatch on the same domain, but different account.

---

### Supporting send-only addresses without a regex table

A recent feature-request was raised for [allowing accounts to opt-out from being considered managed by Postfix](https://github.com/docker-mailserver/docker-mailserver/issues/2813#issuecomment-1272223118). They provide their solution to use another config for `scripts/helpers/postfix.sh:_vhost_collect_postfix_domains()` to exclude any account using a mentioned domain. I believe we have a similar opt-out config feature for when a relayhost is enabled (_due to blanket adding all accounts to the relay.._).

That prompted me as likely a misconfiguration (_**confirmed**, due to lack of our docs covering how to support the use-case_). Their use-case was to send mail as `noreply@<external-domain>`, however `SPOOF_PROTECTION=1` was restricting this without a matching account for that address - but as they wanted to send to a user at that same external domain (_which was now considered managed by Postfix_), it tries to deliver internally instead of outbound to the intended external mail-server. The `noreply` sender address was only for sending to external accounts, not receiving any inbound mail, an account is therefore not necessary to resolve this use-case properly.

I investigated and figured that they'd be better to use an opt-in config for what account should be allowed to send as an external account instead, which the `smtpd_sender_login_maps` can allow just fine. My link to that issue provides an example for how to setup a local testing environment that proves this approach works well. Normally it's not an issue when `SPOOF_PROTECTION=0` (default), as an authenticated user can use whatever sender address they like then.

Only issue is that using regex probably isn't always appropriate - Users need to be careful with what expressions they use, often forgetting to escape regex tokens like `.` when providing a domain (_I came across quite a few snippets where this was the case_). However all the other tables aren't specifically for this feature which could cause issues, so **another table should be made available within the `unionmap`**. I haven't looked over the earlier mentioned [LDAP PR that added a related config ENV](https://github.com/docker-mailserver/docker-mailserver/pull/1902) to extend support, but assume it provides similar flexibility when needed.

If `SPOOF_PROTECTION=1` was to be enabled in future as a default, we should first support configuring with **an alternative table (_eg: `texthash:/etc/postfix/senders-map.cf` or similar would be fine_)**, along with supporting documentation and release notes (_probably warrants a notice period before switching the default for the feature?_).

### External accounts being handled as internal accidentally?

Related is [another scenario / issue](https://github.com/docker-mailserver/docker-mailserver/issues/2673#issuecomment-1272103802), again regarding a `noreply` sending account (_but this one would belong to the internal domain_). It's recipients are provided from LDAP accounts, and some users are remote / external addresses, such as `user@gmail.com`.

I'm inclined to think this is more of an issue with their LDAP config, as DMS should not be treating `@gmail` as managed, unless your config is creating them as internal mail accounts for DMS (_equivalent of `postfix-accounts.cf`_), or misconfigured as aliases (_since we collect any alias domain into the Postfix `/etc/postfix/vhost` to be treated as managed by DMS_).

Their desire was to try match the recipient mail is to be sent out to in hopes of convincing Postfix to deliver as outbound mail. Again, these accounts were not to receive mail to DMS, or forward / relay any inbound mail arriving to DMS to these external mail-servers (eg: Gmail). The only intention was having the `noreply@example.com` send out notifications from a service to these LDAP users associated mail accounts (_where some of the LDAP users are `example.com` accounts, but not all_).

In this case rather than matching the `FROM` envelope, I believe we could workaround by using a `transport_maps` table that matches the `TO` (recipient). The prior send-only issue above did have a solution proposed to opt-out of managing entire domains (_by excluding them from Postfix `/etc/postfix/vhost`_), I'm just not convinced that's something we should maintain.

I lack LDAP knowledge, but cannot think of how this would be setup in a non-LDAP config where it'd make sense to create `@gmail` accounts - but I have seen users create aliases to remote addresses / accounts like Gmail? (_which can introduce the same problem if the remote domain is incorrectly added as the alias, not the recipient that an alias should map to_)

We have the following described in [our docs at the end of this example](https://docker-mailserver.github.io/docker-mailserver/edge/examples/tutorials/basic-installation/):

> _Create email accounts and aliases:_
> 
> **With `SPOOF_PROTECTION=0`**
> 
> ```
> ./setup.sh email add admin@example.com passwd123
> ./setup.sh email add info@example.com passwd123
> ./setup.sh alias add admin@example.com external-account@gmail.com
> ./setup.sh alias add info@example.com external-account@gmail.com
> ./setup.sh email list
> ./setup.sh alias list
> ```
> 
> Aliases make sure that any email that comes to these accounts is forwarded to your third-party email address (`external-account@gmail.com`), where they are retrieved (_eg: via third-party web or mobile app_), instead of connecting directly to `docker-mailserver` with POP3 / IMAP.

Directly after that `SPOOF_PROTECTION=0` example is one for `SPOOF_PROTECTION=1`, where the advice is to create an intermediate alias, to workaround the sender map problem. That works because we include the alias table of aliases mapping to whatever recipients, and treating those as owners. **A workaround that shouldn't be needed / used**.

The [Postfix virtual docs on alias forwarding](https://www.postfix.org/VIRTUAL_README.html#forwarding) has a similar example (_excluding the `SPOOF_PROTECTION` issue_).

---

###  `setup alias add`

The `setup alias add` command "usage" / help could better match our `addalias` utility which is more clear on the inputs?:

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/bin/setup#L46-L47

The internal utility makes it more apparent the first input is an alias that is mapped to recipient(s) (_internal `postfix-accounts.cf`, external addresses, or other aliases_):

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/bin/addalias#L21-L22

- `setup alias add alias@example.com account@example.com` will create an alias `alias@example.com` that maps to internal account `account@example.com`. `setup email list` will show that alias associated to the account as expected.
- `setup alias add user@example.com user@gmail.com` will create an alias `user@example.com` that will map to a remote `user@gmail.com` account. Mail received / submitted to `user@example.com` is forwarded to that remote address instead of stored in a local mailbox. **Unless** Postfix believes it manages that domain.
- **BUG:** An alias can be created that is identical to an existing account, and vice versa with account creation for existing aliases. We probably shouldn't allow that? `setup email list` if `user@example.com` exists as an account also, will list the `user@example.com` alias as an alias to the account itself.. but that's a side-effect bug from allowing such in the first place.

Apart from the bug issue, [our aliases docs](https://docker-mailserver.github.io/docker-mailserver/edge/config/user-management/aliases/) communicate roughly the same two scenarios described above for `setup alias add` :+1: 

### Alias domains being added to `/etc/postfix/vhost`

We've had this support [introduced way back in Oct 2015](https://github.com/docker-mailserver/docker-mailserver/pull/26), which now lives in a [dedicated vhost helper script](https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/scripts/helpers/postfix.sh#L60-L71) that [I created during a refactoring via this PR](https://github.com/docker-mailserver/docker-mailserver/pull/2620) (June 2022).

I discussed further details regarding it and related Postfix `main.cf` config in the [send-only linked issue above](https://github.com/docker-mailserver/docker-mailserver/issues/2813#issuecomment-1272394563). As the `postfix-regexp.cf` for alias support came after that Oct 2015 PR, and our scripts can't realistically scrape domains from a regex, it could be better to have Postfix use the correct tables to lookup directly when it needs to, which would allow for the alias regex.

Postfix provides separate config to check for accounts and aliases regarding if it should be considered the final destination for delivery of mail received/sent. We could better leverage that support directly. It would also allow for users to more easily opt-out domains I think, should they really need to.

**Comment by georglauterbach** [comment]
Nice! We could make people that want to start working on DMS guide to this issue as a nice "beginner-friendly" way of getting to know the project.

**Comment by polarathene** [comment]
Potentially useful notes related to this topic when revisiting this issue.

---

Recently `SPOOF_PROTECTION=1` being changed to default for the `v12` release was proposed.

This issue was [cited to discourage that](https://github.com/docker-mailserver/docker-mailserver/pull/3127#issuecomment-1451160023) for the time being due to variety of config changes already introduced into `v12` and the concerns raised around the `SPOOF_PROTECTION` feature support in this issue above.

---

This may be useful to migrate into technical documentation for maintainers in future.

The `SPOOF_PROTECTION=1` feature is implemented in Postfix with `smtpd_sender_restrictions` with any of the rejection policies listed below, presently the feature chooses `reject_authenticated_sender_login_mismatch`.

`smtpd_sender_login_maps` example config can be found [here in Postfix docs](https://www.postfix.org/SASL_README.html#server_sasl_authz_envelope) in addition to usage with `main.cf:smtpd_sender_restrictions` (_docs incorrectly assign to `smtpd_recipient_restrictions`_). The [_Additional Notes_ section](https://github.com/docker-mailserver/docker-mailserver/issues/2819#issue-1402114383) above provides more info on the `smtpd_sender_login_maps` config format and our usage of `regex` + `unionmap` config tables with it.

- [`smtpd_sender_login_maps`](https://www.postfix.org/postconf.5.html#smtpd_sender_login_maps) - Config for SASL account lookups used by settings these settings when SASL is enabled:
  - [`reject_authenticated_sender_login_mismatch`](https://www.postfix.org/postconf.5.html#reject_authenticated_sender_login_mismatch)
    - Can only send via addresses permitted for their account.
  - [`reject_known_sender_login_mismatch`](https://www.postfix.org/postconf.5.html#reject_known_sender_login_mismatch)
    - Not permitted to send as any existing mail account (_unless belonging to this user_), otherwise permitted.
    - This is a more relaxed variant of `reject_authenticated_sender_login_mismatch`, only protecting spoofing of known accounts.
  - [`reject_unauthenticated_sender_login_mismatch`](https://www.postfix.org/postconf.5.html#reject_unauthenticated_sender_login_mismatch)
    - Not permitted to send as any existing mail account, otherwise permitted.
    - Only applies when SASL is enabled, but the mail client sending has not authenticated via SASL (_eg: permitted to send via network whitelist, without authenticating_).
  - [`reject_sender_login_mismatch`](https://www.postfix.org/postconf.5.html#reject_sender_login_mismatch)
    - Alias that applies both `reject_authenticated_sender_login_mismatch` and `reject_unauthenticated_sender_login_mismatch`.

---

`SPOOF_PROTECTION` feature code has since migrated to here (_with upcoming changes in [this PR](https://github.com/docker-mailserver/docker-mailserver/pull/3127)_):

https://github.com/docker-mailserver/docker-mailserver/blob/b451742f0a6c23bd68fcc952fd3f6b6f9d291e34/target/scripts/startup/setup.d/security/spoofing.sh#L13-L28

When someone has time to, the conditionals there can be refactored, as described in the **Task Overview** at the start of this issue. LDAP perhaps can be updated later if easier.

**Comment by toto6038** [comment]
I noticed that the [script](https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/scripts/startup/setup.d/security/spoofing.sh) doesn't properly set `smtpd_sender_restrictions` to [`$mua_sender_restrictions`](https://github.com/docker-mailserver/docker-mailserver/blob/a3571a88c135f7f6a627d82d312fbae6b1c7b817/target/postfix/main.cf#L75) when `SPOOF_PROTECTION=1`. 
Just want to clarify if it's leaved on intention or a miss? Would be glad to open PR to address it.

**Comment by polarathene** [comment]
> I noticed that the [script](https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/scripts/startup/setup.d/security/spoofing.sh) doesn't properly set `smtpd_sender_restrictions` to [`$mua_sender_restrictions`](https://github.com/docker-mailserver/docker-mailserver/blob/a3571a88c135f7f6a627d82d312fbae6b1c7b817/target/postfix/main.cf#L75) when `SPOOF_PROTECTION=1`.

It doesn't need to as that is already the case (_[More info here](https://github.com/docker-mailserver/docker-mailserver/pull/4379#issue-2864959395)_):

https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/postfix/master.cf#L26

https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/postfix/master.cf#L40

---

The `=0` branch modifies the parameter to remove a restriction enabled by default in our `main.cf`. It's a bit backwards since `=0` is the default, so I can understand how that can look confusing.

https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/scripts/startup/setup.d/security/spoofing.sh#L26

https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/postfix/main.cf#L70-L75

**Comment by toto6038** [comment]
@polarathene Thanks for your explanation! As the `smtpd_sender_restrictions` setting will eventually be overridden to `$mua_sender_restrictions` in master.cf. Maybe we can consider commenting out this line to reduce confusion?
https://github.com/docker-mailserver/docker-mailserver/blob/23bb1c8e50dad1462c645b8a9cf50aeee8bc2625/target/postfix/main.cf#L59

**Comment by polarathene** [comment]
As I detailed in [this link](https://github.com/docker-mailserver/docker-mailserver/pull/4379#issue-2864959395) from my prior comment, no `smtpd` covers port 25, not just the MUA submission(s) ports (465 + 587). Those are different requirements.

**Comment by toto6038** [comment]
@polarathene Oh I finally understand the context and the usage of `dms_smtpd_sender_restrictions`. I find my confusion stemmed from unfamiliarity to how DMS separates the intended usage of port 25, 456, and 587. Thank you for elaborating it!
